{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "58878efb-43a8-4bc5-a986-0608e86ead95",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to teach students how to run and interpret metagenome read recruitment onto SAGs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6af40f8-b03d-406e-bc1b-a23d00e9b605",
   "metadata": {},
   "source": [
    "## Data Environment Setup\n",
    "\n",
    "We will be downloading (or copying over) metagenome data from NCBI to use for this project.\n",
    "\n",
    "I've done this in advance of the lesson, but I want you all to see how it is done for future projects.\n",
    "\n",
    "First we'll create a project directory for this lesson.  \n",
    "Then, we'll create a data directory in storage.\n",
    "Finally, we'll link our data directory to our project directory.  The code will look something like this:\n",
    "\n",
    "I have made a directory for metagenomic data within our storage directory:  \n",
    "```\n",
    "$ mkdir -p /mnt/storage/data/day3AM_metagenomes/\n",
    "```\n",
    "\n",
    "Make a project directory:\n",
    "```\n",
    "$ mkdir ~/storage/user_lab/{your_username}/metagenome_recruitment_lesson\n",
    "```\n",
    "\n",
    "Now navigate into your project directory ( ~/storage/user_lab/{your_username}/metagenome_recruitment_lesson) and start a jupyter notebook using the file system navigator side bar.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55574be2-92b9-4217-8cdd-731eb7f031f3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# make a directory for data within your storage directory:\n",
    "!mkdir -p ~/storage/user_lab/julia/metagenome_recruitment/data/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba500a85-3e41-4060-8c1b-f662e3c39a5f",
   "metadata": {},
   "source": [
    "## Software Environment Setup\n",
    "\n",
    "The software needed for this tutorial can be found in the conda env config file ./coverm_env.yml\n",
    "\n",
    "It is pre-installed on the jupyter hub.  To activate it in the terminal type:\n",
    "\n",
    "```\n",
    "conda activate coverM\n",
    "```\n",
    "\n",
    "**TODO: this is currently a local environment for me... I need to install it as root so that everyone has access.**\n",
    "\n",
    "To use it in a jupyter notebook, select the 'coverm' kernel on the upper right-hand side of the screen."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce8557d7-8055-4937-9e3c-53979503315f",
   "metadata": {},
   "source": [
    "## Data Setup\n",
    "\n",
    "For this lesson, we are going to use the SAGs that we have been working with all week, AG-910, to examine their abundance over time at BATS using metagenomic read recruitment.  \n",
    "\n",
    "Here's the data we will access:\n",
    "\n",
    "We will download metagenomic reads from BATS from this bioproject: https://www.ncbi.nlm.nih.gov/bioproject?term=PRJNA385855\n",
    "\n",
    "I've downloaded a metadata sheet with all sra metagenomes from this bioproject to: ./data/PRJNA385855_sra_metadata.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e70110c3-e849-4b74-8a2f-9d000c654330",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Run', 'Assay Type', 'AvgSpotLen', 'Bases', 'BioProject', 'BioSample',\n",
       "       'BioSampleModel', 'bottle_id', 'Bytes', 'Center Name',\n",
       "       'Collection_date', 'Consent', 'cruise_id', 'DATASTORE filetype',\n",
       "       'DATASTORE provider', 'DATASTORE region', 'Depth', 'env_biome',\n",
       "       'env_feature', 'env_material', 'Experiment', 'geo_loc_name_country',\n",
       "       'geo_loc_name_country_continent', 'geo_loc_name', 'Instrument',\n",
       "       'isolation-source', 'lat_lon', 'Library Name', 'LibraryLayout',\n",
       "       'LibrarySelection', 'LibrarySource', 'Organism', 'Platform',\n",
       "       'ReleaseDate', 'Sample Name', 'SRA Study'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"./data/PRJNA385855_sra_metadata.csv\", sep = \",\")\n",
    "\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ada9cdc-bc6c-41cd-a44e-8d3d007283d4",
   "metadata": {},
   "source": [
    "Let's look at the metagenomes within this dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c93cbf74-6fe1-4853-a484-af5b0e316408",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>Collection_date</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>Depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR6507277</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2009-08-19</td>\n",
       "      <td>HOT214</td>\n",
       "      <td>5m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR6507278</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2009-11-04</td>\n",
       "      <td>HOT216</td>\n",
       "      <td>100m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR6507279</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2009-07-14</td>\n",
       "      <td>BATS248</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR6507280</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2009-11-07</td>\n",
       "      <td>BATS252</td>\n",
       "      <td>100m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR5720219</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-10-10</td>\n",
       "      <td>HOT164</td>\n",
       "      <td>5m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR5720221</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-08-15</td>\n",
       "      <td>HOT162</td>\n",
       "      <td>175m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR5720223</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-09-28</td>\n",
       "      <td>HOT163</td>\n",
       "      <td>115m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR5720224</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-09-28</td>\n",
       "      <td>HOT163</td>\n",
       "      <td>175m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR5720226</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-06-15</td>\n",
       "      <td>HOT160</td>\n",
       "      <td>175m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR5720228</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-08-15</td>\n",
       "      <td>HOT162</td>\n",
       "      <td>100m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR5720230</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2003-07-15</td>\n",
       "      <td>BATS178</td>\n",
       "      <td>120m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR5720232</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2003-02-21</td>\n",
       "      <td>BATS173</td>\n",
       "      <td>40m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR5720234</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-12-08</td>\n",
       "      <td>BATS195</td>\n",
       "      <td>180m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR5720236</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2003-03-22</td>\n",
       "      <td>BATS174</td>\n",
       "      <td>180m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR5720238</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2003-03-22</td>\n",
       "      <td>BATS174</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR5720240</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2003-01-17</td>\n",
       "      <td>HOT144</td>\n",
       "      <td>125m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRR5720242</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-10-10</td>\n",
       "      <td>HOT164</td>\n",
       "      <td>175m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR5720244</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-11-10</td>\n",
       "      <td>HOT165</td>\n",
       "      <td>100m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR5720246</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-12-14</td>\n",
       "      <td>HOT166</td>\n",
       "      <td>5m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR5720249</td>\n",
       "      <td>WGS</td>\n",
       "      <td>2004-04-21</td>\n",
       "      <td>BATS187</td>\n",
       "      <td>40m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run Assay Type Collection_date cruise_id Depth\n",
       "0   SRR6507277        WGS      2009-08-19    HOT214    5m\n",
       "1   SRR6507278        WGS      2009-11-04    HOT216  100m\n",
       "2   SRR6507279        WGS      2009-07-14   BATS248   10m\n",
       "3   SRR6507280        WGS      2009-11-07   BATS252  100m\n",
       "4   SRR5720219        WGS      2004-10-10    HOT164    5m\n",
       "5   SRR5720221        WGS      2004-08-15    HOT162  175m\n",
       "6   SRR5720223        WGS      2004-09-28    HOT163  115m\n",
       "7   SRR5720224        WGS      2004-09-28    HOT163  175m\n",
       "8   SRR5720226        WGS      2004-06-15    HOT160  175m\n",
       "9   SRR5720228        WGS      2004-08-15    HOT162  100m\n",
       "10  SRR5720230        WGS      2003-07-15   BATS178  120m\n",
       "11  SRR5720232        WGS      2003-02-21   BATS173   40m\n",
       "12  SRR5720234        WGS      2004-12-08   BATS195  180m\n",
       "13  SRR5720236        WGS      2003-03-22   BATS174  180m\n",
       "14  SRR5720238        WGS      2003-03-22   BATS174    1m\n",
       "15  SRR5720240        WGS      2003-01-17    HOT144  125m\n",
       "16  SRR5720242        WGS      2004-10-10    HOT164  175m\n",
       "17  SRR5720244        WGS      2004-11-10    HOT165  100m\n",
       "18  SRR5720246        WGS      2004-12-14    HOT166    5m\n",
       "19  SRR5720249        WGS      2004-04-21   BATS187   40m"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Run','Assay Type','Collection_date','cruise_id','Depth']].head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4629227-ff50-425f-b745-bb4698d26a7f",
   "metadata": {},
   "source": [
    "Our SAG plate is from BATS248.  We are interested in examining the abundance of SAGs in the surface at BATS within this metagenome collection, so let's find just those metagenomes collected at BATS from either 1m or 10m depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d836e635-35a5-4f17-9083-50792b61df8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21 metagenomes that we are interested in downloading\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Collection_date</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>SRR5720233</td>\n",
       "      <td>2003-02-21</td>\n",
       "      <td>BATS173</td>\n",
       "      <td>SAMN07137079</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR5720238</td>\n",
       "      <td>2003-03-22</td>\n",
       "      <td>BATS174</td>\n",
       "      <td>SAMN07137082</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>SRR5720327</td>\n",
       "      <td>2003-04-22</td>\n",
       "      <td>BATS175</td>\n",
       "      <td>SAMN07137064</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SRR5720283</td>\n",
       "      <td>2003-05-20</td>\n",
       "      <td>BATS176</td>\n",
       "      <td>SAMN07137103</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>SRR5720235</td>\n",
       "      <td>2003-07-15</td>\n",
       "      <td>BATS178</td>\n",
       "      <td>SAMN07137085</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SRR5720286</td>\n",
       "      <td>2003-08-12</td>\n",
       "      <td>BATS179</td>\n",
       "      <td>SAMN07137088</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>SRR5720332</td>\n",
       "      <td>2003-10-07</td>\n",
       "      <td>BATS181</td>\n",
       "      <td>SAMN07137067</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>SRR5720276</td>\n",
       "      <td>2003-11-04</td>\n",
       "      <td>BATS182</td>\n",
       "      <td>SAMN07137106</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>SRR5720262</td>\n",
       "      <td>2003-12-02</td>\n",
       "      <td>BATS183</td>\n",
       "      <td>SAMN07137109</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>SRR5720338</td>\n",
       "      <td>2004-01-27</td>\n",
       "      <td>BATS184</td>\n",
       "      <td>SAMN07137070</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>SRR5720322</td>\n",
       "      <td>2004-02-24</td>\n",
       "      <td>BATS185</td>\n",
       "      <td>SAMN07137121</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>SRR5720337</td>\n",
       "      <td>2004-03-23</td>\n",
       "      <td>BATS186</td>\n",
       "      <td>SAMN07137073</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR5720256</td>\n",
       "      <td>2004-04-21</td>\n",
       "      <td>BATS187</td>\n",
       "      <td>SAMN07137091</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SRR5720257</td>\n",
       "      <td>2004-05-18</td>\n",
       "      <td>BATS188</td>\n",
       "      <td>SAMN07137112</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>SRR5720260</td>\n",
       "      <td>2004-06-15</td>\n",
       "      <td>BATS189</td>\n",
       "      <td>SAMN07137115</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>SRR5720321</td>\n",
       "      <td>2004-08-17</td>\n",
       "      <td>BATS191</td>\n",
       "      <td>SAMN07137118</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SRR5720251</td>\n",
       "      <td>2004-09-14</td>\n",
       "      <td>BATS192</td>\n",
       "      <td>SAMN07137094</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>SRR5720307</td>\n",
       "      <td>2004-10-13</td>\n",
       "      <td>BATS193</td>\n",
       "      <td>SAMN07137097</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>SRR5720278</td>\n",
       "      <td>2004-11-12</td>\n",
       "      <td>BATS194</td>\n",
       "      <td>SAMN07137100</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>SRR5720342</td>\n",
       "      <td>2004-12-08</td>\n",
       "      <td>BATS195</td>\n",
       "      <td>SAMN07137076</td>\n",
       "      <td>1m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR6507279</td>\n",
       "      <td>2009-07-14</td>\n",
       "      <td>BATS248</td>\n",
       "      <td>SAMN08390924</td>\n",
       "      <td>10m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Run Collection_date cruise_id     BioSample Depth\n",
       "74   SRR5720233      2003-02-21   BATS173  SAMN07137079    1m\n",
       "14   SRR5720238      2003-03-22   BATS174  SAMN07137082    1m\n",
       "119  SRR5720327      2003-04-22   BATS175  SAMN07137064   10m\n",
       "99   SRR5720283      2003-05-20   BATS176  SAMN07137103    1m\n",
       "75   SRR5720235      2003-07-15   BATS178  SAMN07137085   10m\n",
       "38   SRR5720286      2003-08-12   BATS179  SAMN07137088   10m\n",
       "64   SRR5720332      2003-10-07   BATS181  SAMN07137067    1m\n",
       "96   SRR5720276      2003-11-04   BATS182  SAMN07137106    1m\n",
       "90   SRR5720262      2003-12-02   BATS183  SAMN07137109    1m\n",
       "124  SRR5720338      2004-01-27   BATS184  SAMN07137070    1m\n",
       "117  SRR5720322      2004-02-24   BATS185  SAMN07137121    1m\n",
       "66   SRR5720337      2004-03-23   BATS186  SAMN07137073    1m\n",
       "22   SRR5720256      2004-04-21   BATS187  SAMN07137091    1m\n",
       "87   SRR5720257      2004-05-18   BATS188  SAMN07137112   10m\n",
       "89   SRR5720260      2004-06-15   BATS189  SAMN07137115   10m\n",
       "57   SRR5720321      2004-08-17   BATS191  SAMN07137118    1m\n",
       "83   SRR5720251      2004-09-14   BATS192  SAMN07137094    1m\n",
       "49   SRR5720307      2004-10-13   BATS193  SAMN07137097    1m\n",
       "97   SRR5720278      2004-11-12   BATS194  SAMN07137100    1m\n",
       "127  SRR5720342      2004-12-08   BATS195  SAMN07137076    1m\n",
       "2    SRR6507279      2009-07-14   BATS248  SAMN08390924   10m"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dataframe of metagenomes of interest\n",
    "mgoi = df[df['cruise_id'].str.contains('BATS') & df['Depth'].isin(['10m','1m'])][['Run','Collection_date','cruise_id','BioSample','Depth']].sort_values(by = 'Collection_date')\n",
    "\n",
    "# going to save this table to file\n",
    "mgoi.to_csv(\"./data/bats_metagenomes_of_interest.csv\", index=False)\n",
    "\n",
    "print(\"There are\", len(mgoi), \"metagenomes that we are interested in downloading\")\n",
    "mgoi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30a07e46-3ccb-4d9a-852c-84f7e2a3524b",
   "metadata": {},
   "source": [
    "Just for ease of access, I'm going to print out another small text file that has just the runids of the metagenomes we are interested in downloading into the directory we will download the files into:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7926d803-0d04-4376-a3a9-6010e5c50de6",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/jupyter-julia/storage/user_lab/julia/metagenome_recruitment/data/metagenomes_to_download.txt', 'w') as oh:\n",
    "    for run in mgoi['Run']:\n",
    "        print(run, file = oh)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f198f67-ed68-4fc8-be0c-aa01a2187124",
   "metadata": {},
   "source": [
    "I'm next going to download these metagenomes from ncbi using the fastq-dump function from the sra-tools package.  Some parameters I'll use, I'm going to skip technical reads, and download only 1000000 reads from each metagenome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01617c25-d23c-4a2a-bd20-b557e15041de",
   "metadata": {},
   "source": [
    "Open up a temrinal window and then run:\n",
    "\n",
    "```\n",
    "conda activate coverm\n",
    "\n",
    "cd ~/storage/user_lab/julia/metagenome_recruitment/data/\n",
    "\n",
    "while read p; do\n",
    "  fastq-dump --split-files --skip-technical -N 0 -X 1000000 --gzip --readids \"$p\"\n",
    "done <metagenomes_to_download.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "543649c9-452a-488d-8dc9-2e8ab7b86678",
   "metadata": {},
   "source": [
    "Let's check which metagenomes were downloaded, using the python package 'glob' which identifies files using wildcards, and returns them as a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8cf5ddbb-4b09-4208-8235-0f63da98ff90",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "\n",
    "mgs = glob('/home/jupyter-julia/storage/user_lab/julia/metagenome_recruitment/data/*_1.fastq.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d7f1829-19ad-42d7-9af9-3a03260d98bd",
   "metadata": {},
   "source": [
    "Now I'm going to use what's called a 'list comprehension' to extract the runids from this list of files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ebc512eb-3ed0-45cb-b7b3-b600ada2106c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "runids = [i.split(\"/\")[-1].split(\"_\")[0] for i in mgs]\n",
    "len(runids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f6424b-05ea-4981-9750-c0748d2d2297",
   "metadata": {},
   "source": [
    "Hmmmm... two SRA archives are missing, I wonder which ones those are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a3737d78-a4df-4062-91b5-e21f1e7270be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Collection_date</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Depth</th>\n",
       "      <th>downloaded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>SRR5720233</td>\n",
       "      <td>2003-02-21</td>\n",
       "      <td>BATS173</td>\n",
       "      <td>SAMN07137079</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR5720238</td>\n",
       "      <td>2003-03-22</td>\n",
       "      <td>BATS174</td>\n",
       "      <td>SAMN07137082</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>SRR5720327</td>\n",
       "      <td>2003-04-22</td>\n",
       "      <td>BATS175</td>\n",
       "      <td>SAMN07137064</td>\n",
       "      <td>10m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SRR5720283</td>\n",
       "      <td>2003-05-20</td>\n",
       "      <td>BATS176</td>\n",
       "      <td>SAMN07137103</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>SRR5720235</td>\n",
       "      <td>2003-07-15</td>\n",
       "      <td>BATS178</td>\n",
       "      <td>SAMN07137085</td>\n",
       "      <td>10m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SRR5720286</td>\n",
       "      <td>2003-08-12</td>\n",
       "      <td>BATS179</td>\n",
       "      <td>SAMN07137088</td>\n",
       "      <td>10m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>SRR5720332</td>\n",
       "      <td>2003-10-07</td>\n",
       "      <td>BATS181</td>\n",
       "      <td>SAMN07137067</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>SRR5720276</td>\n",
       "      <td>2003-11-04</td>\n",
       "      <td>BATS182</td>\n",
       "      <td>SAMN07137106</td>\n",
       "      <td>1m</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>SRR5720262</td>\n",
       "      <td>2003-12-02</td>\n",
       "      <td>BATS183</td>\n",
       "      <td>SAMN07137109</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>SRR5720338</td>\n",
       "      <td>2004-01-27</td>\n",
       "      <td>BATS184</td>\n",
       "      <td>SAMN07137070</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>SRR5720322</td>\n",
       "      <td>2004-02-24</td>\n",
       "      <td>BATS185</td>\n",
       "      <td>SAMN07137121</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>SRR5720337</td>\n",
       "      <td>2004-03-23</td>\n",
       "      <td>BATS186</td>\n",
       "      <td>SAMN07137073</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR5720256</td>\n",
       "      <td>2004-04-21</td>\n",
       "      <td>BATS187</td>\n",
       "      <td>SAMN07137091</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SRR5720257</td>\n",
       "      <td>2004-05-18</td>\n",
       "      <td>BATS188</td>\n",
       "      <td>SAMN07137112</td>\n",
       "      <td>10m</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>SRR5720260</td>\n",
       "      <td>2004-06-15</td>\n",
       "      <td>BATS189</td>\n",
       "      <td>SAMN07137115</td>\n",
       "      <td>10m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>SRR5720321</td>\n",
       "      <td>2004-08-17</td>\n",
       "      <td>BATS191</td>\n",
       "      <td>SAMN07137118</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SRR5720251</td>\n",
       "      <td>2004-09-14</td>\n",
       "      <td>BATS192</td>\n",
       "      <td>SAMN07137094</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>SRR5720307</td>\n",
       "      <td>2004-10-13</td>\n",
       "      <td>BATS193</td>\n",
       "      <td>SAMN07137097</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>SRR5720278</td>\n",
       "      <td>2004-11-12</td>\n",
       "      <td>BATS194</td>\n",
       "      <td>SAMN07137100</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>SRR5720342</td>\n",
       "      <td>2004-12-08</td>\n",
       "      <td>BATS195</td>\n",
       "      <td>SAMN07137076</td>\n",
       "      <td>1m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR6507279</td>\n",
       "      <td>2009-07-14</td>\n",
       "      <td>BATS248</td>\n",
       "      <td>SAMN08390924</td>\n",
       "      <td>10m</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Run Collection_date cruise_id     BioSample Depth downloaded\n",
       "74   SRR5720233      2003-02-21   BATS173  SAMN07137079    1m        yes\n",
       "14   SRR5720238      2003-03-22   BATS174  SAMN07137082    1m        yes\n",
       "119  SRR5720327      2003-04-22   BATS175  SAMN07137064   10m        yes\n",
       "99   SRR5720283      2003-05-20   BATS176  SAMN07137103    1m        yes\n",
       "75   SRR5720235      2003-07-15   BATS178  SAMN07137085   10m        yes\n",
       "38   SRR5720286      2003-08-12   BATS179  SAMN07137088   10m        yes\n",
       "64   SRR5720332      2003-10-07   BATS181  SAMN07137067    1m        yes\n",
       "96   SRR5720276      2003-11-04   BATS182  SAMN07137106    1m         no\n",
       "90   SRR5720262      2003-12-02   BATS183  SAMN07137109    1m        yes\n",
       "124  SRR5720338      2004-01-27   BATS184  SAMN07137070    1m        yes\n",
       "117  SRR5720322      2004-02-24   BATS185  SAMN07137121    1m        yes\n",
       "66   SRR5720337      2004-03-23   BATS186  SAMN07137073    1m        yes\n",
       "22   SRR5720256      2004-04-21   BATS187  SAMN07137091    1m        yes\n",
       "87   SRR5720257      2004-05-18   BATS188  SAMN07137112   10m         no\n",
       "89   SRR5720260      2004-06-15   BATS189  SAMN07137115   10m        yes\n",
       "57   SRR5720321      2004-08-17   BATS191  SAMN07137118    1m        yes\n",
       "83   SRR5720251      2004-09-14   BATS192  SAMN07137094    1m        yes\n",
       "49   SRR5720307      2004-10-13   BATS193  SAMN07137097    1m        yes\n",
       "97   SRR5720278      2004-11-12   BATS194  SAMN07137100    1m        yes\n",
       "127  SRR5720342      2004-12-08   BATS195  SAMN07137076    1m        yes\n",
       "2    SRR6507279      2009-07-14   BATS248  SAMN08390924   10m        yes"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgoi.loc[~mgoi['Run'].isin(runids),'downloaded'] = 'no'\n",
    "mgoi['downloaded'] = mgoi['downloaded'].fillna('yes')\n",
    "mgoi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6cfde1fa-a680-4936-9770-0b3284bad498",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgoi.to_csv(\"./data/bats_metagenomes_of_interest.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9da7a1bc-e1f1-48c9-9771-c7b9192f1b2d",
   "metadata": {},
   "source": [
    "The error I see in the terminal says:\n",
    "\n",
    "```\n",
    "2022-03-23T19:51:04 fastq-dump.2.8.0 err: no error - error with http open 'https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos2/sra-pub-run-11/SRR5720276/SRR5720276.1'\n",
    "2022-03-23T19:51:04 fastq-dump.2.8.0 err: item not found while constructing within virtual database module - the path 'SRR5720276' cannot be opened as database or table\n",
    "```\n",
    "\n",
    "For whatever reason, these two metagenomes can't be downloaded. I'm OK with that, we'll have enough information with the metagenomes we recovered to look into SAG abundance at BATS over time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0345d0fe-e4d7-438e-bcd1-30ccce9251fa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
